src/plugins/hidsim/qmp_connection.cpp

```diff
@@ -0,0 +1,236 @@
+{
+    int ret = -1;
+
+    /* Creates a socket */
+    qc->fd = socket(AF_UNIX, SOCK_STREAM, 0);
+
+    if (qc->fd == -1)
+    {
+        fprintf(stderr, "socket: %s\n", strerror(errno));
+        return qc->fd;
+    }
+    qc->sa.sun_family = AF_UNIX;
+
+    /* Check, if path to domain is too long */
+    if ((strlen(path) + 1) > sizeof(qc->sa.sun_path))
+    {
+        fprintf(stderr, "path to domain socket exceeds %ld characters\n", sizeof(qc->sa.sun_path) );
+        return qc->fd;
+    }
+
+    /* Sets socket path */
+    strcpy(qc->sa.sun_path, path);
+
+    /* Initializes connection */
+    ret = connect(qc->fd, (struct sockaddr*) &qc->sa, sizeof(qc->sa));
+
+    if (ret == -1)
+    {
+        fprintf(stderr, "connect: %s - %s\n", strerror(errno), qc->sa.sun_path);
+        return -1;
+    }
+
+    /* Reads the capabilities, which are sent by the qmp-server */
+    ret = read(qc->fd, qc->buf, sizeof(qc->buf));
+
+    /* Stores capabilities for later use */
+    qc->capabilities = json_tokener_parse(qc->buf);
+    memset(qc->buf, 0, sizeof(qc->buf));
+
+    /* Exits capabilities negotiation mode */
+    ret = write(qc->fd, QMP_EXEC_CAPABILITIES, strlen(QMP_EXEC_CAPABILITIES));
+
+    if (ret == -1)
+    {
+        fprintf(stderr, "write to %s - %s\n", qc->sa.sun_path, strerror(errno));
+        return -1;
+    }
+    /* Check return value */
+    ret = read(qc->fd, qc->buf, sizeof(qc->buf));
+
+    if (ret == -1)
+    {
+        fprintf(stderr, "read from %s - %s\n", qc->sa.sun_path, strerror(errno));
+        return -1;
+    }
+
+    ret = strcmp(qc->buf, QMP_SUCCESS);
+    if (ret != 0)
+    {
+        fprintf(stderr, "Capability negotiation failed, msg: %s\n", qc->buf);
+        return -1;
+    }
+
+    qc->is_connected = true;
+
+    return 0;
+}
+
+/* Sends the given data to the QMPConnection and reads the response */
+int qmp_communicate_json(qmp_connection* qc, struct json_object* in, struct json_object** out)
+{
+
+    const char* in_str = json_object_to_json_string(in);
+    char* out_str = NULL;
+    qmp_communicate(qc, in_str, &out_str);
+    *out = json_tokener_parse(out_str);
+
+    if (out_str)
+        free(out_str);
+
+    return 0;
+}
+
+/* Sends the given data to the QMPConnection and reads the response */
+int qmp_communicate(qmp_connection* qc, const char* in, char** out)
+{
+
+    int ret = write(qc->fd, in, strlen(in));
+    if (ret == -1)
+    {
+        fprintf(stderr, "write to %s - %s\n", qc->sa.sun_path, strerror(errno));
+        return ret;
+    }
+
+    /* Reads the response */
+    ret = read(qc->fd, qc->buf, sizeof(qc->buf));
+
+    if (ret == -1)
+    {
+        fprintf(stderr, "read from %s - %s\n", qc->sa.sun_path, strerror(errno));
+        return ret;
+    }
+    if (out)
+        *out = strdup(qc->buf);
+
+    memset(qc->buf, 0, sizeof(qc->buf));
+
+    return 0;
+}
+
+/* Clean-up */
+int qmp_close_conn(qmp_connection* qc)
+{
+    close(qc->fd);
+
+    if (qc->capabilities)
```

Comment:
No need to check for null

Reply:
It's dynamically allocated, as one can't be sure, that it is not deleted in the meantime, I think, it is safer to check for NULL before freeing, as a caller, who allocated the `qmp_connection` in question can do, what he wants with the capabilities-ptr.

Reply:
>I think, it is safer to check for NULL

Not really, free will be a noop when called with NULL :wink:

Reply:
Oh yes,....thought this would be undefined behaviour, but you're absolutely right..sorry.

Reply:
Changed...set `qc->capabilities` to NULL, to make it obvious, that this intends to prevent double free-issues